xray/aws_test.go

```diff
@@ -184,7 +184,7 @@ func testClientFailedConnection(ctx context.Context, td *TestDaemon, t *testing.
 	}
 	assert.True(t, subseg.Fault)
 	// Should contain 'marshal' and 'attempt' subsegments only.
-	assert.Len(t, subseg.Subsegments, 2)
+	assert.Len(t, subseg.Subsegments, 9)
```

Comment:
It is increased to 9 because we are now counties retries as seperate subsegments.

`[{"trace_id":"1-638d8bbd-3d48187e1cb3b6e20d3c4cf6","id":"362829c0467a53ee","name":"marshal","start_time":1670220733.9923363,"end_time":1670220733.9923422,"parent_id":"7f2240b611bcbd2f","Dummy":false} {"trace_id":"1-638d8bbd-3d48187e1cb3b6e20d3c4cf6","id":"3e4a1f67d4cd2723","name":"attempt","start_time":1670220733.9923518,"end_time":1670220733.9984694,"parent_id":"7f2240b611bcbd2f","fault":true,"cause":{"working_directory":"/home/runner/work/aws-xray-sdk-go/aws-xray-sdk-go/src/github.com/aws/aws-xray-sdk-go/xray","exceptions":[{"id":"e9cd83bebd6b6cb5","type":"awserr.baseError","message":"RequestError: send request failed\ncaused by: Get \"[https://lambda.fake-moon-1.amazonaws.com/2015-03-31/functions/\](https://lambda.fake-moon-1.amazonaws.com/2015-03-31/functions//)": dial tcp: lookup lambda.fake-moon-1.amazonaws.com: no such host","stack":[{"path":"github.com/aws/aws-xray-sdk-go/xray/aws.go","line":139,"label":"glob..func7"},{"path":"github.com/aws/aws-sdk-go@v1.44.114/aws/request/handlers.go","line":267,"label":"(*HandlerList).Run"},{"path":"github.com/aws/aws-sdk-go@v1.44.114/aws/request/request.go","line":558,"label":"(*Request).Send"},{"path":"github.com/aws/aws-sdk-go@v1.44.114/service/lambda/api.go","line":4204,"label":"(*Lambda).ListFunctionsWithContext"},{"path":"github.com/aws/aws-xray-sdk-go/xray/aws_test.go","line":172,"label":"testClientFailedConnection"},{"path":"github.com/aws/aws-xray-sdk-go/xray/aws_test.go","line":88,"label":"TestAWS.func5.1"},{"path":"testing/testing.go","line":1259,"label":"tRunner"},{"path":"runtime/asm_amd64.s","line":1581,"label":"goexit"}]}]},"subsegments":[{"trace_id":"1-638d8bbd-3d48187e1cb3b6e20d3c4cf6","id":"2958cad1474d4fa5","name":"connect","start_time":1670220733.9924157,"end_time":1670220733.9984667,"parent_id":"7f2240b611bcbd2f","subsegments":[{"trace_id":"1-638d8bbd-3d48187e1cb3b6e20d3c4cf6","id":"f1493163d67a463f","name":"dns","start_time":1670220733.9924548,"end_time":1670220733.9983866,"parent_id":"7f2240b611bcbd2f","fault":true,"cause":{"working_directory":"/home/runner/work/aws-xray-sdk-go/aws-xray-sdk-go/src/github.com/aws/aws-xray-sdk-go/xray","exceptions":[{"id":"c19521928dda5ac7","type":"net.DNSError","message":"lookup lambda.fake-moon-1.amazonaws.com: no such host","stack":[{"path":"github.com/aws/aws-xray-sdk-go/xray/httptrace.go","line":232,"label":"NewClientTrace.func3"},{"path":"net/http/httptrace/trace.go","line":58,"label":"WithClientTrace.func2"},{"path":"net/lookup.go","line":328,"label":"(*Resolver).lookupIPAddr"},{"path":"net/ipsock.go","line":288,"label":"(*Resolver).internetAddrList"},{"path":"net/dial.go","line":221,"label":"(*Resolver).resolveAddrList"},{"path":"net/dial.go","line":406,"label":"(*Dialer).DialContext"},{"path":"net/http/transport.go","line":1166,"label":"(*Transport).dial"},{"path":"net/http/transport.go","line":1604,"label":"(*Transport).dialConn"},{"path":"net/http/transport.go","line":1446,"label":"(*Transport).dialConnFor"},{"path":"runtime/asm_amd64.s","line":1581,"label":"goexit"}]}]},"metadata":{"http":{"dns":{"addresses":[],"coalesced":false}}},"Dummy":false}],"Dummy":false}],"Dummy":false} {"trace_id":"1-638d8bbd-3d48187e1cb3b6e20d3c4cf6","id":"13a7e5a21b225b4d","name":"wait","start_time":1670220733.9985044,"end_time":1670220734.03072,"parent_id":"7f2240b611bcbd2f","Dummy":false} {"trace_id":"1-638d8bbd-3d48187e1cb3b6e20d3c4cf6","id":"2d19b1b7f1953f27","name":"attempt","start_time":1670220734.0307543,"end_time":1670220734.037802,"parent_id":"7f2240b611bcbd2f","fault":true,"cause":{"working_directory":"/home/runner/work/aws-xray-sdk-go/aws-xray-sdk-go/src/github.com/aws/aws-xray-sdk-go/xray","exceptions":[{"id":"500cc5cf223e5fbe","type":"awserr.baseError","message":"RequestError: send request failed\ncaused by: Get \"[https://lambda.fake-moon-1.amazonaws.com/2015-03-31/functions/\](https://lambda.fake-moon-1.amazonaws.com/2015-03-31/functions//)": dial tcp: lookup lambda.fake-moon-1.amazonaws.com: no such host","stack":[{"path":"github.com/aws/aws-xray-sdk-go/xray/aws.go","line":139,"label":"glob..func7"},{"path":"github.com/aws/aws-sdk-go@v1.44.114/aws/request/handlers.go","line":267,"label":"(*HandlerList).Run"},{"path":"github.com/aws/aws-sdk-go@v1.44.114/aws/request/request.go","line":558,"label":"(*Request).Send"},{"path":"github.com/aws/aws-sdk-go@v1.44.114/service/lambda/api.go","line":4204,"label":"(*Lambda).ListFunctionsWithContext"},{"path":"github.com/aws/aws-xray-sdk-go/xray/aws_test.go","line":172,"label":"testClientFailedConnection"},{"path":"github.com/aws/aws-xray-sdk-go/xray/aws_test.go","line":88,"label":"TestAWS.func5.1"},{"path":"testing/testing.go","line":1259,"label":"tRunner"},{"path":"runtime/asm_amd64.s","line":1581,"label":"goexit"}]}]},"subsegments":[{"trace_id":"1-638d8bbd-3d48187e1cb3b6e20d3c4cf6","id":"fdc338e5b7305fbe","name":"connect","start_time":1670220734.030872,"end_time":1670220734.0377984,"parent_id":"7f2240b611bcbd2f","subsegments":[{"trace_id":"1-638d8bbd-3d48187e1cb3b6e20d3c4cf6","id":"d59ea3a57fdc09eb","name":"dns","start_time":1670220734.0309749,"end_time":1670220734.03766,"parent_id":"7f2240b611bcbd2f","fault":true,"cause":{"working_directory":"/home/runner/work/aws-xray-sdk-go/aws-xray-sdk-go/src/github.com/aws/aws-xray-sdk-go/xray","exceptions":[{"id":"17d30764547859eb","type":"net.DNSError","message":"lookup lambda.fake-moon-1.amazonaws.com: no such host","stack":[{"path":"github.com/aws/aws-xray-sdk-go/xray/httptrace.go","line":232,"label":"NewClientTrace.func3"},{"path":"reflect/value.go","line":556,"label":"Value.call"},{"path":"reflect/value.go","line":339,"label":"Value.Call"},{"path":"net/http/httptrace/trace.go","line":203,"label":"(*ClientTrace).compose.func1"},{"path":"net/http/httptrace/trace.go","line":58,"label":"WithClientTrace.func2"},{"path":"net/lookup.go","line":328,"label":"(*Resolver).lookupIPAddr"},{"path":"net/ipsock.go","line":288,"label":"(*Resolver).internetAddrList"},{"path":"net/dial.go","line":221,"label":"(*Resolver).resolveAddrList"},{"path":"net/dial.go","line":406,"label":"(*Dialer).DialContext"},{"path":"net/http/transport.go","line":1166,"label":"(*Transport).dial"},{"path":"net/http/transport.go","line":1604,"label":"(*Transport).dialConn"},{"path":"net/http/transport.go","line":1446,"label":"(*Transport).dialConnFor"},{"path":"runtime/asm_amd64.s","line":1581,"label":"goexit"}]}]},"metadata":{"http":{"dns":{"addresses":[],"coalesced":false}}},"Dummy":false}],"Dummy":false}],"Dummy":false} {"trace_id":"1-638d8bbd-3d48187e1cb3b6e20d3c4cf6","id":"d23d6e6a428d52a9","name":"wait","start_time":1670220734.0378492,"end_time":1670220734.1211703,"parent_id":"7f2240b611bcbd2f","Dummy":false} {"trace_id":"1-638d8bbd-3d48187e1cb3b6e20d3c4cf6","id":"b9438fc09e48153c","name":"attempt","start_time":1670220734.1212044,"end_time":1670220734.1374228,"parent_id":"7f2240b611bcbd2f","fault":true,"cause":{"working_directory":"/home/runner/work/aws-xray-sdk-go/aws-xray-sdk-go/src/github.com/aws/aws-xray-sdk-go/xray","exceptions":[{"id":"20acf8fc4d3c8144","type":"awserr.baseError","message":"RequestError: send request failed\ncaused by: Get \"[https://lambda.fake-moon-1.amazonaws.com/2015-03-31/functions/\](https://lambda.fake-moon-1.amazonaws.com/2015-03-31/functions//)": dial tcp: lookup lambda.fake-moon-1.amazonaws.com: no such host","stack":[{"path":"github.com/aws/aws-xray-sdk-go/xray/aws.go","line":139,"label":"glob..func7"},{"path":"github.com/aws/aws-sdk-go@v1.44.114/aws/request/handlers.go","line":267,"label":"(*HandlerList).Run"},{"path":"github.com/aws/aws-sdk-go@v1.44.114/aws/request/request.go","line":558,"label":"(*Request).Send"},{"path":"github.com/aws/aws-sdk-go@v1.44.114/service/lambda/api.go","line":4204,"label":"(*Lambda).ListFunctionsWithContext"},{"path":"github.com/aws/aws-xray-sdk-go/xray/aws_test.go","line":172,"label":"testClientFailedConnection"},{"path":"github.com/aws/aws-xray-sdk-go/xray/aws_test.go","line":88,"label":"TestAWS.func5.1"},{"path":"testing/testing.go","line":1259,"label":"tRunner"},{"path":"runtime/asm_amd64.s","line":1581,"label":"goexit"}]}]},"subsegments":[{"trace_id":"1-638d8bbd-3d48187e1cb3b6e20d3c4cf6","id":"3dd0e7db9a5f5ef4","name":"connect","start_time":1670220734.12132,"end_time":1670220734.1374173,"parent_id":"7f2240b611bcbd2f","subsegments":[{"trace_id":"1-638d8bbd-3d48187e1cb3b6e20d3c4cf6","id":"19896f19f039f01e","name":"dns","start_time":1670220734.1214204,"end_time":1670220734.1372924,"parent_id":"7f2240b611bcbd2f","fault":true,"cause":{"working_directory":"/home/runner/work/aws-xray-sdk-go/aws-xray-sdk-go/src/github.com/aws/aws-xray-sdk-go/xray","exceptions":[{"id":"d59062dd822fbe86","type":"net.DNSError","message":"lookup lambda.fake-moon-1.amazonaws.com: no such host","stack":[{"path":"github.com/aws/aws-xray-sdk-go/xray/httptrace.go","line":232,"label":"NewClientTrace.func3"},{"path":"reflect/value.go","line":556,"label":"Value.call"},{"path":"reflect/value.go","line":339,"label":"Value.Call"},{"path":"net/http/httptrace/trace.go","line":203,"label":"(*ClientTrace).compose.func1"},{"path":"net/http/httptrace/trace.go","line":58,"label":"WithClientTrace.func2"},{"path":"net/lookup.go","line":328,"label":"(*Resolver).lookupIPAddr"},{"path":"net/ipsock.go","line":288,"label":"(*Resolver).internetAddrList"},{"path":"net/dial.go","line":221,"label":"(*Resolver).resolveAddrList"},{"path":"net/dial.go","line":406,"label":"(*Dialer).DialContext"},{"path":"net/http/transport.go","line":1166,"label":"(*Transport).dial"},{"path":"net/http/transport.go","line":1604,"label":"(*Transport).dialConn"},{"path":"net/http/transport.go","line":1446,"label":"(*Transport).dialConnFor"},{"path":"runtime/asm_amd64.s","line":1581,"label":"goexit"}]}]},"metadata":{"http":{"dns":{"addresses":[],"coalesced":false}}},"Dummy":false}],"Dummy":false}],"Dummy":false} {"trace_id":"1-638d8bbd-3d48187e1cb3b6e20d3c4cf6","id":"6d6f5f40f60b60c1","name":"wait","start_time":1670220734.1374652,"end_time":1670220734.284856,"parent_id":"7f2240b611bcbd2f","Dummy":false} {"trace_id":"1-638d8bbd-3d48187e1cb3b6e20d3c4cf6","id":"77d32a623df9c480","name":"attempt","start_time":1670220734.2848933,"end_time":1670220734.2970057,"parent_id":"7f2240b611bcbd2f","fault":true,"cause":{"working_directory":"/home/runner/work/aws-xray-sdk-go/aws-xray-sdk-go/src/github.com/aws/aws-xray-sdk-go/xray","exceptions":[{"id":"0c32e8b3285323a0","type":"awserr.baseError","message":"RequestError: send request failed\ncaused by: Get \"[https://lambda.fake-moon-1.amazonaws.com/2015-03-31/functions/\](https://lambda.fake-moon-1.amazonaws.com/2015-03-31/functions//)": dial tcp: lookup lambda.fake-moon-1.amazonaws.com: no such host","stack":[{"path":"github.com/aws/aws-xray-sdk-go/xray/aws.go","line":139,"label":"glob..func7"},{"path":"github.com/aws/aws-sdk-go@v1.44.114/aws/request/handlers.go","line":267,"label":"(*HandlerList).Run"},{"path":"github.com/aws/aws-sdk-go@v1.44.114/aws/request/request.go","line":558,"label":"(*Request).Send"},{"path":"github.com/aws/aws-sdk-go@v1.44.114/service/lambda/api.go","line":4204,"label":"(*Lambda).ListFunctionsWithContext"},{"path":"github.com/aws/aws-xray-sdk-go/xray/aws_test.go","line":172,"label":"testClientFailedConnection"},{"path":"github.com/aws/aws-xray-sdk-go/xray/aws_test.go","line":88,"label":"TestAWS.func5.1"},{"path":"testing/testing.go","line":1259,"label":"tRunner"},{"path":"runtime/asm_amd64.s","line":1581,"label":"goexit"}]}]},"subsegments":[{"trace_id":"1-638d8bbd-3d48187e1cb3b6e20d3c4cf6","id":"9bf350d92af26c6e","name":"connect","start_time":1670220734.285034,"end_time":1670220734.2970014,"parent_id":"7f2240b611bcbd2f","subsegments":[{"trace_id":"1-638d8bbd-3d48187e1cb3b6e20d3c4cf6","id":"c3fd6ce4a94a9045","name":"dns","start_time":1670220734.2851355,"end_time":1670220734.296849,"parent_id":"7f2240b611bcbd2f","fault":true,"cause":{"working_directory":"/home/runner/work/aws-xray-sdk-go/aws-xray-sdk-go/src/github.com/aws/aws-xray-sdk-go/xray","exceptions":[{"id":"0301548e2a76d5a8","type":"net.DNSError","message":"lookup lambda.fake-moon-1.amazonaws.com: no such host","stack":[{"path":"github.com/aws/aws-xray-sdk-go/xray/httptrace.go","line":232,"label":"NewClientTrace.func3"},{"path":"reflect/value.go","line":556,"label":"Value.call"},{"path":"reflect/value.go","line":339,"label":"Value.Call"},{"path":"net/http/httptrace/trace.go","line":203,"label":"(*ClientTrace).compose.func1"},{"path":"net/http/httptrace/trace.go","line":58,"label":"WithClientTrace.func2"},{"path":"net/lookup.go","line":328,"label":"(*Resolver).lookupIPAddr"},{"path":"net/ipsock.go","line":288,"label":"(*Resolver).internetAddrList"},{"path":"net/dial.go","line":221,"label":"(*Resolver).resolveAddrList"},{"path":"net/dial.go","line":406,"label":"(*Dialer).DialContext"},{"path":"net/http/transport.go","line":1166,"label":"(*Transport).dial"},{"path":"net/http/transport.go","line":1604,"label":"(*Transport).dialConn"},{"path":"net/http/transport.go","line":1446,"label":"(*Transport).dialConnFor"},{"path":"runtime/asm_amd64.s","line":1581,"label":"goexit"}]}]},"metadata":{"http":{"dns":{"addresses":[],"coalesced":false}}},"Dummy":false}],"Dummy":false}],"Dummy":false} {"trace_id":"1-638d8bbd-3d48187e1cb3b6e20d3c4cf6","id":"5c02cb5d130fc856","name":"wait","start_time":1670220734.2970483,"end_time":1670220734.2970574,"parent_id":"7f2240b611bcbd2f","fault":true,"cause":{"working_directory":"/home/runner/work/aws-xray-sdk-go/aws-xray-sdk-go/src/github.com/aws/aws-xray-sdk-go/xray","exceptions":[{"id":"377d5fada018e9e9","type":"awserr.baseError","message":"RequestError: send request failed\ncaused by: Get \"[https://lambda.fake-moon-1.amazonaws.com/2015-03-31/functions/\](https://lambda.fake-moon-1.amazonaws.com/2015-03-31/functions//)": dial tcp: lookup lambda.fake-moon-1.amazonaws.com: no such host","stack":[{"path":"github.com/aws/aws-xray-sdk-go/xray/aws.go","line":149,"label":"glob..func8"},{"path":"github.com/aws/aws-sdk-go@v1.44.114/aws/request/handlers.go","line":267,"label":"(*HandlerList).Run"},{"path":"github.com/aws/aws-sdk-go@v1.44.114/aws/request/request.go","line":559,"label":"(*Request).Send"},{"path":"github.com/aws/aws-sdk-go@v1.44.114/service/lambda/api.go","line":4204,"label":"(*Lambda).ListFunctionsWithContext"},{"path":"github.com/aws/aws-xray-sdk-go/xray/aws_test.go","line":172,"label":"testClientFailedConnection"},{"path":"github.com/aws/aws-xray-sdk-go/xray/aws_test.go","line":88,"label":"TestAWS.func5.1"},{"path":"testing/testing.go","line":1259,"label":"tRunner"},{"path":"runtime/asm_amd64.s","line":1581,"label":"goexit"}]}]},"Dummy":false}]`